File: src/main/java/de/pflugradts/passbird/application/ExchangeAdapterPort.java
package de.pflugradts.passbird.application;

import de.pflugradts.passbird.domain.model.Tuple;
import de.pflugradts.passbird.domain.model.transfer.Bytes;

import java.util.stream.Stream;

/**
 * AdapterPort for exchanging password data with a 3rd party.
 */
public interface ExchangeAdapterPort {
    void send(Stream<Tuple<Bytes, Bytes>> data);
    Stream<Tuple<Bytes, Bytes>> receive();
}

File: src/main/kotlin/de/pflugradts/passbird/application/ExchangeAdapterPort.kt
package de.pflugradts.passbird.application

import de.pflugradts.kotlinextensions.TryResult
import de.pflugradts.passbird.domain.model.nest.Nest
import de.pflugradts.passbird.domain.model.shell.Shell
import de.pflugradts.passbird.domain.model.shell.ShellPair

interface ExchangeAdapterPort {
    fun send(data: PasswordInfoMap, password: CharArray): TryResult<Unit>
    fun receive(password: CharArray): TryResult<PasswordInfoMap>
}

data class PasswordYolkInfo(
    val secret: Shell,
    val algorithm: String,
    val digits: Int,
    val periodSeconds: Int,
)

data class PasswordInfo(
    val first: ShellPair,
    val second: List<ShellPair>,
    val yolk: PasswordYolkInfo? = null,
)

typealias PasswordInfoMap = Map<Nest, List<PasswordInfo>>
